### PR TITLE
fix(typescript): not allow empty body 'if' with semicolon #1067

### PR DESCRIPTION
**Pull request by samir-hamud**
Fix Issue #1067

### CODE REVIEW

**Comment by samir-hamud** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

File: src/quick-lint-js/fe/parse-statement.cpp [hunk 1]
```diff
@@ -3977,6 +3977,11 @@ void Parser::parse_and_visit_if(Parse_Visitor_Base &v) {
     parse_and_visit_body();
     break;
 
+  case Token_Type::semicolon:
+    if (!this->options_.typescript) {
+      parse_and_visit_body();
+      break;
+    }
```
Comment (strager) [inline src/quick-lint-js/fe/parse-statement.cpp:3984]: Please explicitly fall through:
```suggestion
  case Token_Type::semicolon:
    if (!this->options_.typescript) {
      parse_and_visit_body();
      break;
    }
    [[fallthrough]];
```

But I think it's better copy-paste the code below and invert the `if`:
```suggestion
  case Token_Type::semicolon:
    if (this->options_.typescript) {
      this->diag_reporter_->report(Diag_Missing_Body_For_If_Statement{
          .expected_body =
              Source_Code_Span::unit(this->lexer_.end_of_previous_token()),
      });
    } else {
      parse_and_visit_body();
    }
    break;
```

File: src/quick-lint-js/fe/parse-statement.cpp [hunk 2]
```diff
@@ -3977,6 +3977,11 @@ void Parser::parse_and_visit_if(Parse_Visitor_Base &v) {
     parse_and_visit_body();
     break;
 
+  case Token_Type::semicolon:
```
Comment (strager) [inline src/quick-lint-js/fe/parse-statement.cpp:3980]: Nitpick: It's helpful to include a comment explaining what code this code path covers. For example:
```suggestion
  // if (cond);  // Invalid TypeScript.
  case Token_Type::semicolon:
```

File: test/test-parse-statement.cpp [hunk 3]
```diff
@@ -1244,6 +1244,61 @@ TEST_F(Test_Parse_Statement, disallow_label_named_yield_in_generator_function) {
                             "visit_exit_function_scope",
                         }));
 }
+
+TEST_F(Test_Parse_Statement, if_body_with_semicolon_typescript) {
+  {
+    Spy_Visitor p = test_parse_and_visit_statement(
+        u8"if (a)\n; else e;"_sv,  //
+        u8"      ` Diag_Missing_Body_For_If_Statement"_diag,
```
Comment (strager) [inline test/test-parse-statement.cpp:1252]: Minor: I think the span is a bit confusing. If the user writes a body at the suggested point, they will end up with code like this:
```javascript
if (a) {
}
; else e;
```

I think we should instead highlight the `;` and hint that the `;` needs to be replaced with a body:
```javascript
if (a) {
} else e;
```

```suggestion
        u8"if (a)\n; else e;"_sv,  //
        u8"        ^ Diag_Missing_Body_For_If_Statement"_diag,
```

File: test/test-parse-statement.cpp [hunk 4]
```diff
@@ -1244,6 +1244,61 @@ TEST_F(Test_Parse_Statement, disallow_label_named_yield_in_generator_function) {
...
+        u8"if (a);\n else e;"_sv,  //
+        u8"      ` Diag_Missing_Body_For_If_Statement"_diag,
+        typescript_options);
+    EXPECT_THAT(p.visits, ElementsAreArray({
+                              "visit_variable_use", //a
+                          }));
+  }  
+
+  {
+    Test_Parser p(u8"{\nif (a);\n} b;"_sv, typescript_options, capture_diags);
```
Comment (strager) [inline test/test-parse-statement.cpp:1270]: What motivated this test case?

File: test/test-parse-statement.cpp [hunk 5]
```diff
@@ -1244,6 +1244,61 @@ TEST_F(Test_Parse_Statement, disallow_label_named_yield_in_generator_function) {
                             "visit_exit_function_scope",
                         }));
 }
+
+TEST_F(Test_Parse_Statement, if_body_with_semicolon_typescript) {
```
Comment (strager) [inline test/test-parse-statement.cpp:1248]: We should have a similar test but for `else`'s body. E.g. `if (a) {} else;`

**Review by strager — APPROVED** [review]

**Review by strager — COMMENTED** [review]
Patch looks good to me overall. Let me know when I should merge.

File: test/test-parse-statement.cpp (hunk 5 above)
Comment (strager) [reply test/test-parse-statement.cpp:1248]: (According to [your comment](https://github.com/quick-lint/quick-lint-js/issues/1067#issuecomment-1712256942), such code should produce no diagnostic.)

File: test/test-parse-statement.cpp (hunk 4 above)
Comment (samir-hamud) [reply test/test-parse-statement.cpp]: No relevant motivations.
Just thought the more test cases the better. Inexperience issues.

Changed for simplier and direct tests

**Review by strager — APPROVED** [review]

**Comment by strager** [comment]
Landed as commit aa4c375452ef6c1184e89d3cb5c1d7af945c0f8a.

**Pull request closed without merging**